### PR TITLE
0.2.x (mesh building) release!

### PR DESCRIPTION
**Pull request by SamuelHall700**
# PolarRoute Pull Request Template

Date: 06/06/2023
Version Number: 0.1.0 -> **0.2.19**
 
## Release Notes:

- Big speed increase when loading non-Mercator data in mesh building [#155](https://github.com/antarctica/PolarRoute/issues/155)
- Added splitting on Vector data [#123](https://github.com/antarctica/SDADT-project/issues/123)
- Added the ability to generate meshes wrapping around the entire world [#116 ](https://github.com/antarctica/SDADT-project/issues/116)
- Added the ability export layers of our mesh to `.tif` files [#113 ](https://github.com/antarctica/SDADT-project/issues/113)
- Added the ability for users to implement there own `DataLoader` without requiring them to edit the existing codebase [#143](https://github.com/antarctica/SDADT-project/issues/143) 
- Overhaul of the docs [#124](https://github.com/antarctica/SDADT-project/issues/124) 
I've attached a zip of the docs here for ease-of-use: 
[PolarRoute_docs.zip](https://github.com/antarctica/PolarRoute/files/11665574/PolarRoute_docs.zip)
- PolarRoute can now accept an existing route and predict its expected travel-time and fuel consumption based on the environmental conditions [#97](https://github.com/antarctica/SDADT-project/issues/97)
- Complete rewrite of the smooting code following the new pseudocode #218
- Validation of the syntax of all configs according to new schema with helpful error messages for any mistakes in configuration #220

# Files Changed

docs:
```
docs/source/conf.py
docs/source/index.rst
docs/source/sections/Code_overview.rst
docs/source/sections/Command_line_interface.rst
docs/source/sections/Configuration/Configuration_overview.rst
docs/source/sections/Configuration/Mesh_construction_config.rst
docs/source/sections/Configuration/Route_planning_config.rst
docs/source/sections/Configuration/Vessel_performance_config.rst
docs/source/sections/Dataloaders/AddingDataloaders.rst
docs/source/sections/Dataloaders/DataLoaderInterface.rst
docs/source/sections/Dataloaders/Factory.rst
docs/source/sections/Dataloaders/overview.rst
docs/source/sections/Dataloaders/scalar/implemented/AMSR.rst
docs/source/sections/Dataloaders/scalar/implemented/BSOSEDepth.rst
docs/source/sections/Dataloaders/scalar/implemented/BSOSESeaIce.rst
docs/source/sections/Dataloaders/scalar/implemented/BinaryGRF.rst
docs/source/sections/Dataloaders/scalar/implemented/GEBCO.rst
docs/source/sections/Dataloaders/scalar/implemented/IceNet.rst
docs/source/sections/Dataloaders/scalar/implemented/ScalarGRF.rst
docs/source/sections/Dataloaders/scalar/index.rst
docs/source/sections/Dataloaders/vector/implemented/ORAS5Current.rst
docs/source/sections/Dataloaders/vector/implemented/SOSE.rst
docs/source/sections/Dataloaders/vector/index.rst
docs/source/sections/Discrete_meshing.rst
docs/source/sections/Examples.rst
docs/source/sections/Installation.rst
docs/source/sections/Mesh_Construction/Mesh_construction_classes.rst
docs/source/sections/Mesh_Construction/Mesh_construction_overview.rst
docs/source/sections/Mesh_Construction/Mesh_validation.rst
docs/source/sections/Outputs.rst
docs/source/sections/Route_calculation.rst
docs/source/sections/Route_optimisation.rst
docs/source/sections/ipython_notebooks.rst
docs/source/sections/running.rst
```
dataloaders:
```
polar_route/dataloaders/factory.py
polar_route/dataloaders/scalar/abstract_scalar.py
polar_route/dataloaders/scalar/amsr.py
polar_route/dataloaders/scalar/baltic_sea_ice.py
polar_route/dataloaders/scalar/bsose_depth.py
polar_route/dataloaders/scalar/bsose_sea_ice.py
polar_route/dataloaders/scalar/gebco.py
polar_route/dataloaders/scalar/modis.py
polar_route/dataloaders/scalar/scalar_csv.py
polar_route/dataloaders/scalar/scalar_grf.py
polar_route/dataloaders/scalar/shape.py
polar_route/dataloaders/vector/abstract_vector.py
polar_route/dataloaders/vector/baltic_current.py
polar_route/dataloaders/vector/era5_wind.py
polar_route/dataloaders/vector/north_sea_current.py
polar_route/dataloaders/vector/oras5_current.py
polar_route/dataloaders/vector/sose.py
polar_route/dataloaders/vector/vector_csv.py
polar_route/dataloaders/vector/vector_grf.py
```
mesh generation:
```
polar_route/mesh_generation/aggregated_cellbox.py
polar_route/mesh_generation/boundary.py
polar_route/mesh_generation/cellbox.py
polar_route/mesh_generation/environment_mesh.py
polar_route/mesh_generation/jgrid_aggregated_cellbox.py
polar_route/mesh_generation/jgrid_cellbox.py
polar_route/mesh_generation/mesh_builder.py
polar_route/mesh_generation/metadata.py
polar_route/mesh_generation/neighbour_graph.py
polar_route/mesh_validation/__init__.py
polar_route/mesh_validation/sampler.py
```
vessel performance:
```
polar_route/vessel_performance/abstract_vessel.py
polar_route/vessel_performance/vessels/SDA.py
polar_route/vessel_performance/vessels/underwater_vessel.py
```
route planning:
```
polar_route/route_calc.py
polar_route/route_planner.py
```
tests:
```
tests/regression_tests/mesh_test_functions.py
tests/regression_tests/pytest.ini
tests/regression_tests/route_test_functions.py
tests/regression_tests/test_mesh.py
tests/regression_tests/test_routes_dijkstra.py
tests/regression_tests/test_routes_smoothed.py
tests/regression_tests/test_vessel.py
tests/testing_strategy.md
tests/unit_tests/resources/format_conf.json
tests/unit_tests/resources/global_grf_normal.json
tests/unit_tests/test_env_mesh.py
tests/unit_tests/test_mesh_builder.py
tests/unit_tests/test_vessel_performance.py
```
misc:
```
requirements.txt
setup.py
polar_route/__init__.py
polar_route/cli.py
polar_route/__init__.py
polar_route/cli.py
```
# Testing

I've run the full regression test suite and attached the output here: 
[full_test.txt](https://github.com/antarctica/PolarRoute/files/11665528/full_test.txt)
Good news, they all pass!

# Checklist

- [x] My code follows [pep8](https://peps.python.org/pep-0008/) style guidelines.  
- [x] I have commented my code, particularly in hard-to-understand areas.  
- [x] I have updated the documentation of the codebase where required.  
- [x] My changes generate no new warnings.   
- [ ] ~~My PR has been made to the `0.2.x` branch (**DO NOT SUBMIT A PR TO MAIN**)~~

tags:
@JimCircadian 
@hjabbot 
@ayat-khairy 
@masthorne 
@gecoombs    
@donyazz

### CODE REVIEW

**Comment by SamuelHall700** [comment]
Note from George to remember to re-build the docs before merging to main :)

**Comment by SamuelHall700** [comment]
Adding some code here to hopefully simplify the install for any reviews of this release. 
This should work to install PolarRoute from a bash shell in Windows

```
# clone repo
git clone --branch 0.2.x https://github.com/antarctica/PolarRoute 

# setup virtual environment 
cd PolarRoute
python -m venv venv
source venv/Scripts/activate

# install dependencies 
pip install -r requirements.txt
pip install pipwin
pipwin install gdal

# install PolarRoute
python setup.py install
```

**Comment by foxm1** [comment]
The optimise_routes method fails when the start or destination waypoint is on a cellbox boundary. It seems that in these cases, the waypoints are removed from the accessibility graph at the outset, so no Dijkstra paths can be constructed, which is not a good solution. To cope with these cases properly, we should uniformly allocate every boundary point to a cell (eg: allocate all points on the bottom and left boundaries of a cell, apart from the bottom right and top left corner points, to that cell, and construct the Dijkstra paths (which will contain additional legs at start and/or end). Then smooth these cases by introducing an extra condition in smoothing to avoid merging midpoint and lastpoint when the lastpoint is the end waypoint on a boundary.

**Comment by gecoombs** [comment]
While working on a new current dataloader I found a bug that seems to break most vector dataloaders other than SOSE (#198). Once a fix is implemented it should be included in this release.

**Comment by Ulvetanna** [comment]
This will be my running issues list that will need to be fixed. I will edit this comment over time to include more as I go along.

**_Installation - Manuals_**
- [ ] Github 'Setup.py install' will not work with the requirements.txt. You instead need to get the installation using a local pip install

**Review by foxm1 — CHANGES_REQUESTED** [review]
- [ ] 1. In SDA_config.json, change MinDepth:-10 to EITHER MinDepth: 10 or MaxElevation: -10, as a negative depth does not make sense. 
- [ ] 2. Make sure that a warning is output if the data provided in the config file is insufficient to generate a mesh of the specified region, ie: if the given region is larger than the range of any of the data files specified. Padding the region with NaN is not what the user intended and should not be hidden from the user.
- [ ] 3. Make a less verbose version of -v for create_mesh. Maybe a -u (a user version of -v) which does not show debugging flags like HET HOM but shows progress information.
- [ ] 4. The route configs contain redundant fields. Trim them out as discussed. 
- [ ] 5. Switch the order of the optimise_routes arguments on the docs page (section 4.3)
- [ ] 6. In section 5.1.2 of docs, correct/improve the description of the splitting condition. The three points that must be stated are (where t is the threshold, lb is the lower bound and ub is the upper bound) : (i) The cell is homogenous (should not be split) if the proportion of data points with values above t% is less than lb%. (ii) The cell is homogenous (should not be split) if the proportion of data points with values above t% is greater than ub%. (iii) The cell is inhomogeneous (should be split) if the proportion of data points with values above t% is between lb% and ub%.
- [ ] 7. Explain that it follows that if ub% is zero and lb% is 1, then the cell will always be split.
- [ ] 8. It would be useful to show how to construct a mesh using GRFs (as Jonny already suggested) so that users can experiment before they have to download data.  Once users are ready to download data, they need guidance about the minimum data they need to build sensible meshes. There is a note in section 6 that says this set should include SIC, thickness, density and vector data are needed, but GEBCO (or other bathymetry) is not mentioned. This should be added. Probably SIC and SOSE or GEBCO are the minimum needed. I'm not convinced thickness or density are really essential, and vector data is probably not essential either.
- [ ] 9. Example config files are definitely needed. A mesh config file and a route config file can be copied and pasted from the docs, not ideal. There is a mesh config file and an SDA config file on the san, but no route config file. Also no folder containing example configs in the repo. Therefore let's construct these.
- [ ] 10. General principle - how can we make the process of mesh construction and vehicle performance addition as straightforward as possible for someone coming in fresh to the project?

File: docs/source/sections/Installation.rst
Comment (SamuelHall700) [reply docs/source/sections/Installation.rst]: @foxm1 
2. There is an issue around this currently set to be developed as part of the 1.1 release of the Mesh Builder https://github.com/antarctica/SDADT-project/issues/150. Would you like this moved to the 0.2.x release of PolarRoute instead?

3. There is an issue around this current set to be developed as part of the 1.1 release of the Mesh Builder https://github.com/antarctica/SDADT-project/issues/157. Would you like this moved to the 0.2.x release of PolarRoute?

4. This will be addressed as part of the 0.3.x release of PolarRoute.

7. A `ub%: 1` and` lb%:  0` would not result in a cellbox always being split, this would result in a cellbox being split in _all cases except_ when **all** datapoints in the cellbox are either above or below a given threshold value.

**Comment by Ulvetanna** [comment]
We are just need:

**Review by Ulvetanna — CHANGES_REQUESTED** [review]
I approve changes once the below are complete. As Maria has given the go-ahead to allow once all her changes are met (which are also in the below), then I have removed her as a reviewer as her changes would have been accepted
- [ ] #220

**Review by Ulvetanna — APPROVED** [review]
Now that pull request for #220 is complete I accept the the pull request #196

**Pull request merged**
